v-next/hardhat/test/internal/builtin-plugins/flatten/task-action.ts

```diff
@@ -0,0 +1,446 @@
+import type { FlattenActionArguments } from "../../../../src/internal/builtin-plugins/flatten/task-action.js";
+
+import assert from "node:assert/strict";
+import fs from "node:fs";
+import { afterEach, beforeEach, describe, it } from "node:test";
+
+import { useFixtureProject } from "@nomicfoundation/hardhat-test-utils";
+
+import { createHardhatRuntimeEnvironment } from "../../../../src/hre.js";
+import { getHardhatVersion } from "../../../../src/internal/utils/package.js";
+
+const originalConsoleLog = console.log;
+const originalConsoleWarn = console.warn;
+
+const consoleLogBuffer: string[] = [];
+const consoleWarnBuffer: string[] = [];
+
+const getConsoleLogOutput = () => consoleLogBuffer.join("\n");
+const getConsoleWarnOutput = () => consoleWarnBuffer.join("\n");
+
+async function getExpectedSol(fileName = "expected.sol") {
+  const expected = fs.readFileSync(fileName, "utf8");
+
+  const hardhatVersion = await getHardhatVersion();
+  return expected.replace("{HARDHAT_VERSION}", hardhatVersion);
+}
+
+function getContractsOrder(flattenedFiles: string) {
+  const CONTRACT_REGEX = /\s*contract(\s+)(\w)/gm;
+  const matches = flattenedFiles.match(CONTRACT_REGEX);
+
+  return (matches ?? []).map((m: string) => m.replace("contract", "").trim());
+}
+
+async function assertFlattenedFilesResult(flattened: string) {
+  const expected = await getExpectedSol();
+
+  assert.equal(flattened, expected);
+}
+
+async function createHRE() {
+  return createHardhatRuntimeEnvironment({}, {}, process.cwd());
+}
+
+describe("flatten/task-action", () => {
+  beforeEach(() => {
+    // Replace console.log and console.warn so we can assert on their output
+    consoleLogBuffer.length = 0;
+    consoleWarnBuffer.length = 0;
+
+    console.log = (...args: unknown[]) => {
+      consoleLogBuffer.push(args.join(" "));
+    };
+    console.warn = (...args: unknown[]) => {
+      consoleWarnBuffer.push(args.join(" "));
+    };
+  });
+
+  afterEach(() => {
+    console.log = originalConsoleLog;
+    console.warn = originalConsoleWarn;
+  });
```

Comment:
Instead of doing this, the task action can accept these params

```
print = console.log,
warn = console.warn
```

which are not exposed to the CLI, nor hooked into it in any way.

Then, to test it, we just need to call the task action directly.

Reply:
If we wanted to do that, here's an example - https://github.com/NomicFoundation/hardhat/blob/00119c33b4636f4580be9df87c0e4f5b654b405d/v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts#L16

We do, however, assert on the returned values already so it might not be as necessary to inspect the entire stdio. It might still be useful to suppress it though.

Reply:
I added two extra params for the log and warn functions. Using ReplOptions seemed a bit overkill for these because it's just output (perfect fit for the console action althought).

Reply:
Oh no, I wasn't suggesting repl options. That's task specific. I just wanted to show you how we set up args that are only exposed for testing.

Reply:
Ahh I see. Thanks !